src/util.c

```diff
@@ -368,6 +368,447 @@ static int darwin_sem_destroy(sem_t *sem)
+    }
+
+    return iStatus;
+}
+
+
+static int freertos_clock_gettime( clockid_t clock_id,
+                   struct timespec * tp )
+{
+    TimeOut_t xCurrentTime = { 0 };
+
+    /* Intermediate variable used to convert TimeOut_t to struct timespec.
+     * Also used to detect overflow issues. It must be unsigned because the
+     * behavior of signed integer overflow is undefined. */
+    uint64_t ullTickCount = 0ULL;
+
+    /* Silence warnings about unused parameters. */
+    ( void ) clock_id;
+
+    /* Get the current tick count and overflow count. vTaskSetTimeOutState()
+     * is used to get these values because they are both static in tasks.c. */
+    vTaskSetTimeOutState( &xCurrentTime );
+
+    /* Adjust the tick count for the number of times a TickType_t has overflowed.
+     * portMAX_DELAY should be the maximum value of a TickType_t. */
+    ullTickCount = ( uint64_t ) ( xCurrentTime.xOverflowCount ) << ( sizeof( TickType_t ) * 8 );
+
+    /* Add the current tick count. */
+    ullTickCount += xCurrentTime.xTimeOnEntering;
+
+    /* Convert ullTickCount to timespec. */
+    UTILS_NanosecondsToTimespec( ( int64_t ) ullTickCount * NANOSECONDS_PER_TICK, tp );
+
+    return 0;
+}
+
+#define clock_gettime freertos_clock_gettime
+
+static int freertos_sem_timedwait( sem_t * sem,
+                   const struct timespec * abstime )
+{
+    int iStatus = 0;
+    TickType_t xDelay = portMAX_DELAY;
+    int iPreviousValue = Atomic_Decrement_u32( ( uint32_t * ) &sem->value );
+
+    if( abstime != NULL )
+    {
+        /* If the provided timespec is invalid, still attempt to take the
+         * semaphore without blocking, per POSIX spec. */
+        if( UTILS_ValidateTimespec( abstime ) == 0 )
+        {
+            xDelay = 0;
+            iStatus = EINVAL;
+        }
+        else
+        {
+            struct timespec xCurrentTime = { 0 };
+
+            /* Get current time */
+            if( clock_gettime( CLOCK_REALTIME, &xCurrentTime ) != 0 )
+            {
+                iStatus = EINVAL;
+            }
+            else
+            {
+                iStatus = UTILS_AbsoluteTimespecToDeltaTicks( abstime, &xCurrentTime, &xDelay );
+            }
+
+            /* If abstime was in the past, still attempt to take the semaphore without
+             * blocking, per POSIX spec. */
+            if( iStatus == ETIMEDOUT )
+            {
+                xDelay = 0;
+            }
+        }
+    }
+
+    /* If previous semaphore value is larger than zero, the thread entering this function call
+     * can take the semaphore without yielding. Else (<=0), calling into FreeRTOS API to yield.
+     */
+    if( iPreviousValue > 0 )
+    {
+        /* Under no circumstance shall the function fail with a timeout if the semaphore can be locked immediately. */
+        iStatus = 0;
+    }
+    else
+    {
+        /* Take the semaphore using the FreeRTOS API. */
+        if( xSemaphoreTake( ( SemaphoreHandle_t ) &sem->sem,
```

Comment:
fun fact, `xSemaphoreTakeFromISR()` doesn't have an `xTicksToWait` arg, and seems to have `try` semantics (though the man page is unclear on this), so it can't be dropped in here.  it can only be used behind `freertos_sem_trywait()`, which you might consider.